### DISCUSSION TITLE
CMake AOT Problem

### DISCUSSION THREAD

**Discussion by calein** [halide/Halide#5336] [Q&A] [answered]
Hi,
I am facing a problem that is similar to the one posted [here](https://stackoverflow.com/questions/60649596/the-visibility-of-halide-library-in-subdirectory-using-cmake-building-system?answertab=votes#tab-top). However because @alexreinking meantioned in his comments 6 month ago that the Halide CMake rules will change in the next release, I am not sure if the solution is the same in my case as I am on the current master branch.

I explicitly try to separate the AOT compilation in a generator and a runner part. The runner should not depend on libHalide as mentioned [here](https://halide-lang.org/tutorials/tutorial_lesson_10_aot_compilation_run.html). My Halide project looks like this:

```
├── gen
│   ├── gen.cpp
│   └── CMakeLists.txt
── run
│   ├── run.cpp
│   └── CMakeLists.txt
```

The **/gen/CMakeLists.txt** looks as follows:

```
cmake_minimum_required(VERSION 3.16)
project(gen)

set(CMAKE_CXX_STANDARD 11)

set(Halide_DIR <path-to-halide>/lib/cmake/Halide)
find_package(Halide REQUIRED)

add_executable(gen.generator gen.cpp)
target_link_libraries(gen.generator PRIVATE Halide::Generator)

add_halide_library(blur FROM gen.generator)
```

The generator in **gen.cpp** is the same as shown [here](https://www.youtube.com/watch?v=UeyWo42_PS8).

When I invoke `cmake . && make` in **/gen** the following files are generated (besides the CMake related files):

```
├── gen
│   ├── blur.h
│   └── blur.o
│   └── blur.runtime.o
│   └── gen.generator
│   └── libblur.a
│   └── libblur.runtime.a
```

The **/run/CMakeLists.txt** looks as follows:

```
cmake_minimum_required(VERSION 3.16)
project(run)

set(CMAKE_CXX_STANDARD 11)

add_executable(run run.cpp)
target_link_libraries(run PRIVATE ${CMAKE_CURRENT_LIST_DIR}/../gen/libblur.a)

target_include_directories(run PRIVATE <path-to-halide>/include)
```

The **run.cpp** only includes **HalideBuffer.h** and **../gen/blur.h**.

When I invoke `cmake . && make` in **/run** I get the following output:

> ```
> -- Configuring done
> -- Generating done
> -- Build files have been written to: <path-to-source>/run
> Scanning dependencies of target run
> [ 50%] Building CXX object CMakeFiles/run.dir/run.cpp.o
> [100%] Linking CXX executable run
> ../gen/libblur.a(blur.o): In function `blur':
> halide_buffer_t.cpp:(.text.blur+0x1409): undefined reference to `halide_error_buffer_argument_is_null'
> halide_buffer_t.cpp:(.text.blur+0x145b): undefined reference to `halide_error_buffer_allocation_too_large'
> halide_buffer_t.cpp:(.text.blur+0x1476): undefined reference to `halide_error_buffer_allocation_too_large'
> halide_buffer_t.cpp:(.text.blur+0x1491): undefined reference to `halide_error_buffer_extents_too_large'
> halide_buffer_t.cpp:(.text.blur+0x14ac): undefined reference to `halide_error_buffer_allocation_too_large'
> halide_buffer_t.cpp:(.text.blur+0x14c4): undefined reference to `halide_error_buffer_allocation_too_large'
> halide_buffer_t.cpp:(.text.blur+0x14df): undefined reference to `halide_error_buffer_extents_too_large'
> halide_buffer_t.cpp:(.text.blur+0x14fb): undefined reference to `halide_error_device_dirty_with_no_device_support'
> halide_buffer_t.cpp:(.text.blur+0x1517): undefined reference to `halide_error_host_is_null'
> halide_buffer_t.cpp:(.text.blur+0x1545): undefined reference to `halide_error_bad_type'
> halide_buffer_t.cpp:(.text.blur+0x1561): undefined reference to `halide_error_bad_dimensions'
> halide_buffer_t.cpp:(.text.blur+0x158c): undefined reference to `halide_error_access_out_of_bounds'
> halide_buffer_t.cpp:(.text.blur+0x15a6): undefined reference to `halide_error_buffer_extents_negative'
> halide_buffer_t.cpp:(.text.blur+0x15d2): undefined reference to `halide_error_access_out_of_bounds'
> halide_buffer_t.cpp:(.text.blur+0x15ef): undefined reference to `halide_error_buffer_extents_negative'
> halide_buffer_t.cpp:(.text.blur+0x1609): undefined reference to `halide_error_buffer_extents_negative'
> halide_buffer_t.cpp:(.text.blur+0x1626): undefined reference to `halide_error_buffer_extents_negative'
> halide_buffer_t.cpp:(.text.blur+0x1660): undefined reference to `halide_error_constraint_violated'
> clang: error: linker command failed with exit code 1 (use -v to see invocation)
> CMakeFiles/run.dir/build.make:103: recipe for target 'run' failed
> make[2]: *** [run] Error 1
> CMakeFiles/Makefile2:94: recipe for target 'CMakeFiles/run.dir/all' failed
> make[1]: *** [CMakeFiles/run.dir/all] Error 2
> Makefile:102: recipe for target 'all' failed
> make: *** [all] Error 2
> ```

I am using Ubuntu 18.04 with LLVM 10.0.0 and build Halide from the master branch. I am a beginner in CMake and Halide and tried to follow [this guide](https://github.com/halide/Halide/blob/master/README_cmake.md) to set up the CMakeLists.

Could you please provide me some help to solve this issue or share a minimum example on how to set up such a separated AOT compilation in CMake?

**Comment by alexreinking** [comment, marked as answer]
The short answer is that you aren't linking to `libblur.runtime.a`.

The long answer is that if you want to split up compilation like this, then you should export / install the library target from the `gen` project and then `find_package` from the `run` project to get those targets. Then you would use `target_link_libraries` to link to the imported _target_ rather than a system-specific path embedded in the CMake file (you should really never do this).

I've been meaning to put together an example of this for multi-toolchain cross-compilation for a while, but I have paper deadlines coming up, so I won't be able to get to it for a couple months.

**Reply by alexreinking** [reply]
You should also not set `Halide_DIR` in the CMakeLists.txt file itself. You should set it at the command line with `cmake ... -DHalide_DIR=/path/to/HalideConfig.cmake ...`. It also looks like you're doing in-tree builds? Don't do that. CMake is meant to be used out-of-tree: `cmake ... -S . -B build` will contain build artifacts in a folder called `build`.

**Comment by calein** [comment]
Thanks for your help and advice Alex (nice talk at cppcon btw :+1:). 

My **/run/CMakeLists.txt** now looks as follows:

```
cmake_minimum_required(VERSION 3.16)
project(run)

set(CMAKE_CXX_STANDARD 11)

find_package(Threads)

add_library(blur.runtime STATIC IMPORTED)
set_target_properties(blur.runtime PROPERTIES
                      IMPORTED_LOCATION ${CMAKE_CURRENT_LIST_DIR}/../gen/build/libblur.runtime.a
                      INTERFACE_LINK_LIBRARIES "${CMAKE_DL_LIBS};pthread"
)

add_library(blur STATIC IMPORTED)
set_target_properties(blur PROPERTIES
                      IMPORTED_LOCATION ${CMAKE_CURRENT_LIST_DIR}/../gen/build/libblur.a 
                      INTERFACE_LINK_LIBRARIES blur.runtime
)

add_executable(run run.cpp)

target_link_libraries(run PRIVATE blur)

target_include_directories(run PRIVATE <path-to-halide>/include)
```

As you already pointed out, using system-specific paths in CMake is propably not best practice, so I removed them and also exported and install the generated libraries as suggested. Here is my solution:

The **/gen/CMakeLists.txt** now looks as follows:

```
cmake_minimum_required(VERSION 3.16)
project(gen)

set(CMAKE_CXX_STANDARD 11)

find_package(Halide REQUIRED)

add_executable(blur.generator gen.cpp)
target_link_libraries(blur.generator PRIVATE Halide::Generator)

add_halide_library(blur FROM blur.generator)

set_target_properties(blur.runtime PROPERTIES
        INTERFACE_LINK_LIBRARIES "${CMAKE_DL_LIBS};pthread"
)

set_target_properties(blur PROPERTIES
        INTERFACE_INCLUDE_DIRECTORIES ${Halide_INCLUDE}
        INTERFACE_LINK_LIBRARIES blur.runtime
)

install(TARGETS blur blur.runtime
        EXPORT blur-targets
        ARCHIVE DESTINATION lib
        LIBRARY DESTINATION lib
        RUNTIME DESTINATION bin
)

install(EXPORT blur-targets
        FILE blur-config.cmake
        DESTINATION lib/cmake/blur
)

install(FILES build/blur.h DESTINATION include)
```

I run the following commands to make the generator:

```
cmake -S . -B build -DCMAKE_INSTALL_PREFIX=$PWD/distrib -DHalide_DIR=<path-to-halide>/lib/cmake/Halide/ -DHalide_INCLUDE=<path-to-halide>/include/
cmake --build ./build --target install

```

The **/run/CMakeLists.txt** now looks as follows:

```
cmake_minimum_required(VERSION 3.16)
project(run)

set(CMAKE_CXX_STANDARD 11)

find_package(blur REQUIRED)

add_executable(run run.cpp)

target_link_libraries(run PRIVATE blur)
target_include_directories(run PRIVATE ${Halide_INCLUDE})
```

I run the following commands to make the runner:

```
cmake -S . -B build -Dblur_DIR=$PWD/../gen/distrib/lib/cmake/blur/ -DHalide_INCLUDE=<path-to-halide>/include/
cmake --build ./build
```
